feat(em-xx): Install and configure Prettier/eslint/stylelint/commitlint/typescript

ModulesUnraveled opened the pull request:
## This PR does the following:
- Installs and configured all of the tools in the title
- Lints and formats all files in the src directory

## This PR does NOT do the following:
- Convert old js files to typescript. I renamed the files so that everything would compile, but I've eslintignored all of the legacy files with comments at the top to update them over time. I wanted to get typescript installed so that any new components I write will follow that methodology, and intend to update the existing files as I touch them during future development. I didn't, however, want to spend a bunch of time upfront updating all of the existing code. If you can think of a better way to do this, I'm all ears!

ModulesUnraveled committed:
feat(em-xx): installed/configured prettier/eslint/stylelint/typescript and formatted files in src

ModulesUnraveled committed:
fix(em-xx): updated references to tsx files

ModulesUnraveled committed:
feat(em-xx): installed react-helmet types

ModulesUnraveled commented on .eslintignore:
```diff
@@ -0,0 +1,6 @@
+node_modules
+bin
+gatsby-config.js
+gatsby-node.js
+.eslintrc.js
+commitlint.config.js
```
I had to add these here to stop getting errors like this:
```
/Users/brian/Websites/4K/emulsify-website/.eslintrc.js
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: .eslintrc.js.
The file must be included in at least one of the projects provided
```

It didn't seem like I should need to add them here though, so if you know how to properly configure this, please let me know!

patrickocoffeyo replied:
This error is being thrown because in your `eslint.tsconfig.json` , you're instructing Typescript to ignore js files outside of the `src` directory. 

Instead of ignoring all js files in the root, I'd recommend opening up that `src/**/*.js` include so it doesn't exclude files in the root: `**/*.js`, I think is what you're looking for, or perhaps even `./*.js`. (here's an example: https://github.com/emulsify-ds/emulsify-cli/blob/develop/tsconfig.eslint.json)

ModulesUnraveled replied:
That makes sense now. Thanks!

ModulesUnraveled committed:
fix(em-xx): updated node version, deleted package-lock and updated yarn.lock

ModulesUnraveled commented:
@patrickocoffeyo I requested your review on this, since you're the most familiar with this type of setup (that I'm aware of). It's basically a PR to setup the tooling for future use (as noted in the description - eslint, in specific, is ignored in a bunch of pre-existing files since I haven't updated them to typescript yet)

patrickocoffeyo commented on .husky/commit-msg:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npx --no-install commitlint --edit "$1"
```
Given that you're already installing commitlint, I'd recommend adding an npm script for commitlint, and then calling it from the locally installed bin.

In **package.json**:
```json
//...
  "scripts": {
    "commitlint": "commitlint"
  }
//...
```

In `.husky/commit-msg`:
```bash
npm run commitlint --edit "$1"
```

Alternatively, you can make husky-specific scripts and manage them with the rest of your scripts like so:
 - husky config: https://github.com/emulsify-ds/emulsify-cli/blob/develop/.husky/commit-msg
 - npm script: https://github.com/emulsify-ds/emulsify-cli/blob/develop/package.json#L19

ModulesUnraveled replied:
These `.husky/*` files are created by husky following their documentation: https://typicode.github.io/husky/#/?id=create-a-hook
And Commitlint has this in their docs: https://github.com/conventional-changelog/commitlint#getting-started

I'm can't answer why they do it this way vs an npm script like you recommend, but it's their "official" way...

patrickocoffeyo replied:
@ModulesUnraveled I recommend using the locally-installed version of commitlint, because it'll lock you into a specific version, and allow you to make sure upgrades are coupled with any needed config changes, etc. It also ensures that all developers are using the same version of commitlint.

ModulesUnraveled replied:
That makes sense to me! I'll update it.

Any idea why they don't have that in the docs?

ModulesUnraveled replied:
Oh, and if we're using yarn for everything should I use that in the husky commands as well? so

```
yarn commitlint --edit "$1"
```

instead of `npm run`?

patrickocoffeyo replied:
> Any idea why they don't have that in the docs?

I'm not sure. It works, and perhaps is the path of least resistance, especially since yarn/npm are slightly different?

> Oh, and if we're using yarn for everything should I use that in the husky commands as well?

Yep! I don't personally have a preference between the two package managers, but if you pick one, you wanna buy into it across the board and avoid using both npm and yarn in the same project.

ModulesUnraveled replied:
👍  Thanks!

patrickocoffeyo commented on .nvmrc:
```diff
@@ -1 +1 @@
-13.13.0
+15
```
I'd recommend pinning this to a minor version. Additionally, it may make sense to use v16, as it's the current version: https://nodejs.org/en/about/releases

ModulesUnraveled replied:
Done

patrickocoffeyo reviewed (changes requested):
Overall this looks great, nice work! I left a couple of small comments, LMK what you think!

I additionally noticed that you removed the npm lock file, and added a yarn lock file. If you're switching to yarn, I'd recommend adjusting your `package.json` scripts so they use `yarn` instead of `npm run`, just for the sake of consistency.

ModulesUnraveled committed:
chore(emul-9): updated eslintignore

ModulesUnraveled committed:
chore(emul-9): updated npm requirement

ModulesUnraveled committed:
chore(emul-9): updated package.json scripts to use yarn for consistency

ModulesUnraveled commented:
Thanks @patrickocoffeyo! It looks like they had both package/yarn lock files in here. I noticed that the CI tool was running yarn, so I deleted the package lock, and updated the yarn one. Good call on updating the npm scripts to be consistent.

I addressed everything here except for the commitlint comment because the way I'm doing it is the way their docs say to. If you feel strongly that it should be done a different way, PLEASE let me know. I don't have any information/knowledge regarding the difference, so I'm literally just leaving it because it's documented that way. BUT if there is a benefit to changing it, I'm all ears!

If this passes all of the tests, I'm going to go ahead and merge to unblock other work, but feel free to comment on the commitlint stuff, and I can address it in another PR.

Thanks again!
